include/aws/common/array_list_with_map.h

```diff
@@ -0,0 +1,82 @@
+#ifndef AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+#define AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/common.h>
+#include <aws/common/hash_table.h>
+
+struct aws_array_list_with_map {
```

Comment:
Hide the implementation details using an opaque impl/state struct similar to hash_table:
https://github.com/awslabs/aws-c-common/blob/d9031cf4f02fe4d8c24a53bb6c8df67ce8be9f0c/include/aws/common/hash_table.h#L48-L51

Reply:
good call, or we can use new/destroy. I don't know which one will be better.